### PR TITLE
[SE-0276] Support multiple patterns in catch clauses

### PR DESCRIPTION
**Pull request by owenv**
e.g. `do { ... } catch MyErr.a(let code), MyErr.b(let code) { ... }`

The approach I took here was to get rid of `CatchStmt` entirely and reuse `CaseStmt` in `DoCatchStmt`. If anyone has time to take a look, I'd appreciate feedback on whether this seems like a good approach before I pitch this on the forums. Initially I tried keeping `CatchStmt` around and just modifying it to hold `CaseLabelItems`, but it resulted in a lot of code duplication in Sema and SILGen. This approach is cleaner overall, but there are still some rough edges (in the profiling code, for instance).

**Requires the following PRs for testing:**

apple/llvm-project#538
apple/swift-syntax#190

### CODE REVIEW

**Comment by owenv** [comment]
Updated to fix code completion tests, this now passes the test suite

**Comment by rintaro** [comment]
I'm a little uncomfortable with making `CaseItemSyntax` have *is_optional* `Pattern`.
Instead of using `CaseStmt` for `catch`, how about making base class (e.g. `AbstractMatchStmt`) from which `CaseStmt` and `CatchStmt` inherit?

**Comment by jrose-apple** [comment]
I'm not sure what benefit we'd get from having CatchStmt and CaseStmt be different classes. They do exactly the same thing, just with different expectations about the incoming type.

**Comment by owenv** [comment]
@jrose-apple There are a few cases where we need to check the `CaseParentKind` to handle cases and catches differently which might cause some concern. Specifically, the Profiler and Formatting both need to special-case their `CaseStmt` handling in a few different places. I think some of the complexity can be cleaned up though, so the tradeoffs are probably worth it unless I discover other problems. 

@rintaro I agree that the `CaseItemSyntax` change is not great. I'll look into ways of cleaning that up a bit when I have some more time to work on this. It might be best to not try to reuse the `case` constructs at the Syntax level.

**Comment by owenv** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by owenv** [comment]
@DougGregor I guess I forgot to link them on this PR,  https://github.com/apple/llvm-project/pull/538 and https://github.com/apple/swift-syntax/pull/190 are also required to build this. Sorry about that!

**Comment by theblixguy** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test
@swift-ci please test source compatibility
@swift-ci please build toolchain

**Comment by theblixguy** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by theblixguy** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test source compatibility

**Comment by owenv** [comment]
Thanks @theblixguy !

**Comment by owenv** [comment]
Let's see if this still passes the test suite...

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by theblixguy** [comment]
It might be worth running a test on Windows too, just in case...

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test windows platform

**Comment by owenv** [comment]
@rintaro do you mind taking another look at these changes when you have a chance? I think I've resolved some of your concerns by differentiating `case`s and `catch`es at the syntax item level.

@gottesmm do you mind taking a look at these SILGen changes or recommending a reviewer? The output SIL looks reasonable, but I'm not very experienced in this area.

File: include/swift/AST/Stmt.h [hunk 1]
```diff
@@ -1072,7 +983,9 @@ class CaseStmt final
   }
   SourceLoc getEndLoc() const { return getBody()->getEndLoc(); }
   SourceRange getLabelItemsRange() const {
-    return ColonLoc.isValid() ? SourceRange(getLoc(), ColonLoc) : getSourceRange();
+    return ItemTerminatorLoc.isValid()
+               ? SourceRange(getLoc(), ItemTerminatorLoc)
+               : getSourceRange();
```
Comment (rintaro) [inline include/swift/AST/Stmt.h]: This includes `{` loc for `catch` statement. Considering the current usages, I don't think that affects anything, but I feel like it shouldn't. But this doesn't block this from merging though.

Comment (owenv) [reply include/swift/AST/Stmt.h]: Good point, I hadn't considered that but it was inconsistent with how ranges work for other statements. I've updated this so the range now extends from `case` to the `:` in switch cases, and from `catch` to the end of the last item in catch clauses, so the `{` is excluded. For `catch {...}`, the range now only covers `catch`.

**Review by rintaro — APPROVED** [review]
Parser and Syntax change LGTM

**Comment by owenv** [comment]
fixed some merge conflicts w/ synthesized comparable for enums

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by owenv** [comment]
@gottesmm friendly ping

**Comment by owenv** [comment]
Updated for recent function builder changes

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please smoke test

**Comment by owenv** [comment]
If anyone has the bandwidth to review the typechecker/silgen changes, I'd really appreciate it!

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please smoke test

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please clean smoke test macOS

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by owenv** [comment]
For reference, the update above picks up changes from recent updates to function builders and the indentation implementation

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test macOS

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test macOS

**Comment by owenv** [comment]
Looks like something in swift-syntax has fallen out of date. I’ll take a closer look tommorrow

**Comment by owenv** [comment]
testing w/ updated swift-syntax PR

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by owenv** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test macOS

**Comment by owenv** [comment]
@rjmccall do you mind helping review the type checker and SILGen changes here as the original author of do-catch?

**Comment by owenv** [comment]
My lldb patch seems to be out of date again, looking now

**Comment by owenv** [comment]
On further investigation, the failures above may have just been infra issues. trying again:

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by owenv** [comment]
fixing profiler merge conflicts

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test

**Comment by owenv** [comment]
macOS failure looks like flakiness:
```
09:52:38 /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.4.sdk/usr/include/objc/objc-api.h:29:10: fatal error: cannot open file '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.4.sdk/usr/include/AvailabilityMacros.h': Operation not permitted
09:52:38 #include <AvailabilityMacros.h>
09:52:38          ^
09:52:38 1 error generated.
```

Windows failure is expected since it doesn't support cross-repo PR testing

apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test macOS platform

File: lib/Sema/TypeCheckStmt.cpp [hunk 2]
```diff
@@ -1258,11 +1237,109 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
...
 
-    // Check all the catch clauses independently.
-    for (auto clause : S->getCatches()) {
-      checkCatchStmt(clause);
+    SmallVector<VarDecl *, 8> scratchMemory1;
+    SmallVector<VarDecl *, 8> scratchMemory2;
+
+    auto clauses = S->getCatches();
+    CaseStmt *previousBlock = nullptr;
+    for (auto i = clauses.begin(), e = clauses.end(); i != e; ++i) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckStmt.cpp:1245]: This code looks nearly identical to what's in `visitSwitchStmt`. Can they be unified?

**Review by DougGregor — APPROVED** [review]
This looks really good! I have one comment where it looks like we have a lot of code duplication in the type checker between `switch` and `do-catch`, but we can clean that up in a follow-up PR.

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
apple/llvm-project#538
apple/swift-syntax#190

@swift-ci please test source compatibility

File: lib/SILGen/SILGenPattern.cpp [hunk 3]
```diff
@@ -2919,38 +2921,145 @@ void SILGenFunction::emitSwitchFallthrough(FallthroughStmt *S) {
...
+    // If we don't have a multi-pattern 'catch', we can emit the
+    // body inline. Emit the statement here and bail early.
+    if (clause->getCaseLabelItems().size() == 1) {
+      // If we have case body vars, set them up to point at the matching var
+      // decls.
+      if (clause->hasCaseBodyVariables()) {
+        // Since we know that we only have one case label item, grab its pattern
+        // vars and use that to update expected with the right SILValue.
+        //
+        // TODO: Do we need a copy here?
```
Comment (slavapestov) [inline lib/SILGen/SILGenPattern.cpp:2946]: @gottesmm Any thoughts on this TODO?

**Review by slavapestov — APPROVED** [review]

**Comment by DougGregor** [comment]
@owenv Also, can you update the ChangeLog in a follow-up PR?

File: lib/Sema/TypeCheckStmt.cpp (hunk 2 above)
Comment (owenv) [reply lib/Sema/TypeCheckStmt.cpp:1245]: I think they can be if I refactor fallthrough handling a little. I'll look into that asap

**Pull request merged**